api/src/app/modules/project/project.service.ts

```diff
@@ -725,5 +750,53 @@ export class ProjectService extends DataService<Project, Repository<Project>, Pr
     this.logger.debug(`${updatedCounts} FOM(s) for ${projectIds} were updated to ${workflowStateCode}`);
   }
 
+  private async updatePublicNoticeIfRequired(projectEntity: Project, stateTransition: WorkflowStateEnum) {
+		if (WorkflowStateEnum.PUBLISHED == stateTransition) {
+			const publicNotices = projectEntity.publicNotices;
+			// set public notice post date to project commenting open date if post date is not set.
+			if (!_.isEmpty(publicNotices) && !publicNotices[0].postDate) {
+				const pnId = publicNotices[0].id;
+				const commentingOpenDate = projectEntity.commentingOpenDate;
+				const updateFields = 
+				{
+					postDate: commentingOpenDate,
+					updateUser: 'system',
+					updateTimestamp: new Date(),
+					revisionCount: () => "revision_count + 1"
+				}
+				const updatedCount = (await this.repository
+					.createQueryBuilder()
+					.update(PublicNotice)
+					.set(updateFields)
+					.where('id =:pnId', {pnId: pnId})
+					.execute()).affected;
+				this.logger.debug(`${updatedCount} FOM (${projectEntity.id}) is updated for public notice (${pnId}) 
+					post_date set to ${projectEntity.commentingOpenDate}`);
+			}
+		}
+  }	
 }
 
+// --- utility validation functions
+
+/**
+ * Public Notice postDate validation: on or before commenting start date.
+ * This is exported validation to be conveniently used in other service.
+ * @param postDate YYYY-MM-DD date string for public notice post date.
+ * @param commentingOpenDate YYYY-MM-DD date string for FOM (Project) commentingOpenDate
+ * @returns true only if postDate on or before commentingOpenDate; 
+ */
+export function isPNPostdateOnOrBeforeCommentingOpenDate(postDate: string, commentingOpenDate: string) {
+	if (_.isEmpty(commentingOpenDate) || !isValidDateOnlyString(commentingOpenDate)) {
+			throw new InternalServerErrorException(`Invalid argument commentingOpenDate: ${commentingOpenDate}`)
```

Comment:
Since this is refactored into DateTimeUtil class, I also change it to throw just Error (we don't currently has custom Error on our end), hope that will make more sense. Currently this is like an util function and does not directly involved in controller returns, could be called from different places, so I gives it an general error. But I would be happy to change it to BadRequestException if you prefer.

Reply:
Thanks Basil, we will keep this as is.